[CBRD-24478]  Make sure that the index is not skewed by duplicate keys.(After Change Spec.)

ctshim opened the pull request:
http://jira.cubrid.org/browse/CBRD-24478

* Reflect changed requirements
1) FK
    system parameter "deduplicate_fk_level": default value is 32, unchangeable in HIDDEN, SESSION
    FK is automatically specified using the "deduplicate_fk_level" value.
    (To create an FK that does not use this function, you must specify "deduplicate_fk_level=0" and record it in the conf file.)
    FK does not show deduplicate information in output in any case.
   
    Exceptionally, the syntax that can define an FK that does not use this function is the ALTER TABLE ADD CONSTRAINT statement. (Private)
    ex) alter table tbl_s ADD CONSTRAINT FOREIGN KEY (id) WITH DEDUPLICATE=0 References tbl_m(id);
   
         
2) NORMAL INDEX
    system parameter "deduplicate_key_level": Default value is 10, can be changed in SESSION, HA delivery
                     "deduplicate_min_keys": Default value is 2, can be changed in SESSION, HA delivery
    2-1) In case of omitting DEDUPLICATE designation
        Automatically set as "deduplicate_key_level" value
        (If you do not want to apply the deduplicate function by default, set "deduplicate_key_level=0" in cubrid.conf.)
        If "deduplicate_min_keys" is not 0, it is applied only when the number of key fields is less than the set value of "deduplicate_min_keys".
          
    2-2) When DEDUPLICATE is specified
       If level is not specified, 32 is assigned (with deduplicate).
       If level is specified, the specified level is used.
      
    In UnloadDB, SHOW CREATE TABLE, and ;SC commands, if the deduplicate_level of the index is 0, the output is not displayed.

3) Avg_num_value_per_deduplicate_key item added (corresponds to Avg_num_value_per_key)
    * fix avg_rec_len calculation

ctshim committed:
fix conflict

ctshim committed:
add PRM_FOR_SERVER property

ctshim committed:
fix show index capacity and diagdb -d4

ctshim committed:
slip, fix  calc average value

ctshim committed:
slip, for only debug

beyondykk9 commented on src/object/class_object.c:
```diff
@@ -8244,6 +8258,30 @@ classobj_check_index_exist (SM_CLASS_CONSTRAINT * constraints, char **out_shared
+	  if (constraint_type == DB_CONSTRAINT_FOREIGN_KEY)
+	    {
+	      int level;
+	      SM_ATTRIBUTE **attp = existing_con->attributes;
```
assert (attp != NULL) is not needed?

ctshim replied:
This is checked in the classobj_find_constraint_by_attrs() function performed earlier.
If existing_con exists, it must have an attribute.

kisoo-han committed:
[CBRD-24824] rpm - set file perm of /etc/profile.d/cubrid.csh to 644 (#4385)

RPM - set file protection mode of /etc/profile.d/cubrid.csh to 644

kisoo-han committed:
libedit - back to emacs profile for csql

shparkcubrid committed:
[CBRD-24813] Problem that term.pred_ord is not initialized. (#4372)

http://jira.cubrid.org/browse/CBRD-24813

init term.pred_ord

ctshim committed:
prevent overflow of bit shift results

beyondykk9 committed:
[CBRD-24828] Executor error when using udf() and bind variable in update query (#4394)

http://jira.cubrid.org/browse/CBRD-24828

beyondykk9 committed:
[CBRD-24828] commented: Executor error when using udf() and bind variable in update query (#4396)

http://jira.cubrid.org/browse/CBRD-24828

Added comment to refer JIRA issue for explaining in detail.

ctshim committed:
adjust pkeys_size to choice covering index

ctshim committed:
Revert "adjust pkeys_size to choice covering index"

This reverts commit 00110fe0007f057cb0188791b5057e588f52c0ab.

youngjinj commented on src/parser/csql_grammar.y:
```diff
@@ -21587,45 +21557,129 @@ opt_encrypt_algorithm
+               $$ = 1;  // thread count is 0
+           DBG_PRINT}}
+	| ONLINE PARALLEL opt_equal unsigned_integer
+	   {{ DBG_TRACE_GRAMMAR(opt_ononline_parallelline, |  ONLINE PARALLEL opt_equal unsigned_integer);
```
```suggestion
	   {{ DBG_TRACE_GRAMMAR(online_parallel, |  ONLINE PARALLEL opt_equalsign unsigned_integer);
```

ctshim replied:
Thank you

youngjinj commented on src/parser/csql_grammar.y:
```diff
@@ -21587,45 +21557,129 @@ opt_encrypt_algorithm
+          }
+        ;        
+
+opt_equal
+        : /* empty */
+        | '='
+        ;
```
Redundant with `opt_equalsign` rule.

ctshim replied:
Thank you, I will fix it.

youngjinj commented on src/parser/csql_grammar.y:
```diff
@@ -21587,45 +21557,129 @@ opt_encrypt_algorithm
+	   {{ DBG_TRACE_GRAMMAR(online_parallel, : ONLINE);
+               $$ = 1;  // thread count is 0
+           DBG_PRINT}}
+	| ONLINE PARALLEL opt_equal unsigned_integer
```
```suggestion
	| ONLINE PARALLEL opt_equalsign unsigned_integer
```

ctshim replied:
OK,

youngjinj commented on src/parser/csql_grammar.y:
```diff
@@ -21587,45 +21557,129 @@ opt_encrypt_algorithm
+          DBG_PRINT}
+        ;
+
+index_with_item_list  
+        : online_parallel
+          { DBG_TRACE_GRAMMAR(index_with_item_list, : online_parallel );
+                container_3 ctn;                
+		SET_CONTAINER_3(ctn, $1, prm_get_integer_value (PRM_ID_DEDUPLICATE_KEY_LEVEL), prm_get_integer_value (PRM_ID_DEDUPLICATE_MIN_KEYS));
+		$$ = ctn;
+          }
+        | online_parallel ',' deduplicate_key_mod_level
+          { DBG_TRACE_GRAMMAR(index_with_item_list, | online_parallel ',' deduplicate_key_mod_level );
+                container_3 ctn;
+		SET_CONTAINER_3(ctn, $1, $3, DEDUPLICATE_MIN_KEYS_UNUSE);
+		$$ = ctn;
+          }
+        | deduplicate_key_mod_level
+          { DBG_TRACE_GRAMMAR(index_with_item_list, | deduplicate_key_mod_level );
+                container_3 ctn;
+		SET_CONTAINER_3(ctn, 0, $1, DEDUPLICATE_MIN_KEYS_UNUSE);
+		$$ = ctn;
+          }
+        | deduplicate_key_mod_level ',' online_parallel
+          { DBG_TRACE_GRAMMAR(index_with_item_list, | deduplicate_key_mod_level ',' online_parallel ); 
+                container_3 ctn;
+		SET_CONTAINER_3(ctn, $3, $1, DEDUPLICATE_MIN_KEYS_UNUSE);
+		$$ = ctn;
+          }
+        ;        
```
Is there any more elegant way?
```
index_with_item_list  
        : online_parallel
        | online_parallel ',' deduplicate_key_mod_level
        | deduplicate_key_mod_level
        | deduplicate_key_mod_level ',' online_parallel
        ;
```
Even if we add only one new option, it should change like below.
```
index_with_item_list  
        : online_parallel
        | online_parallel ',' deduplicate_key_mod_level
        | online_parallel ',' <new_option>
        | online_parallel ',' deduplicate_key_mod_level ',' <new_option>
        | online_parallel ',' <new_option> ',' deduplicate_key_mod_level
        | deduplicate_key_mod_level
        | deduplicate_key_mod_level ',' online_parallel
        | deduplicate_key_mod_level ',' <new_option>
        | deduplicate_key_mod_level ',' online_parallel ',' <new_option>
        | deduplicate_key_mod_level ',' <new_option> ',' online_parallel
        | <new_option>
        | <new_option> ',' online_parallel
        | <new_option> ',' deduplicate_key_mod_level
        | <new_option> ',' online_parallel ',' deduplicate_key_mod_level
        | <new_option> ',' deduplicate_key_mod_level ',' online_parallel
        ;
```
The following code is used in postgresql.
```
alterfunc_opt_list:
                        /* At least one option must be specified */
                        common_func_opt_item
                        | alterfunc_opt_list common_func_opt_item 
                ;

common_func_opt_item:
                        CALLED ON NULL_P INPUT_P
                        | RETURNS NULL_P ON NULL_P INPUT_P
                        | STRICT_P
                        | IMMUTABLE
                        | STABLE
                        | VOLATILE
                        | EXTERNAL SECURITY DEFINER
                        | EXTERNAL SECURITY INVOKER
                        | SECURITY DEFINER
                        | SECURITY INVOKER
                        | LEAKPROOF
                        | NOT LEAKPROOF
                        | COST NumericOnly
                        | ROWS NumericOnly
                        | SUPPORT any_name
                        | FunctionSetResetClause
                        | PARALLEL ColId
                ;

static bool
compute_common_attribute(...)
{
        if (strcmp(defel->defname, "volatility") == 0)
        {
                if (is_procedure)
                        goto procedure_error;
                if (*volatility_item)
                        goto duplicate_error;

                *volatility_item = defel;
        }
        ...
duplicate_error:
        ereport(ERROR,
                        (errcode(ERRCODE_SYNTAX_ERROR),
                         errmsg("conflicting or redundant options"),
                         parser_errposition(pstate, defel->location)));
        return false;
...
```

ctshim replied:
Modifications to be flexible will be carried out as a separate work later.

youngjinj commented on src/parser/csql_grammar.y:
```diff
@@ -21587,45 +21557,129 @@ opt_encrypt_algorithm
+           /* It can be entered through the table constraint clause of "CREATE TABLE" and the add constraint clause of "ALTER TABLE".
+            * If it is entered in the "CREATE TABLE" statement, an error is processed.
+           */          
+            if(parser_is_null_alter_node())
```
Is the purpose of preventing developers from adding this rule elsewhere?

ctshim replied:
For FK, only "with deduplicate=0" can be specified using the ALTER TABLE statement. However, it is not a feature that is open to the general user.For FK, only "with deduplicate=0" can be specified using the ALTER TABLE statement. However, it is not a feature that is open to the general user.

youngjinj replied:
I was mistaken that this rule is not used for create_stmt. Thank you.

beyondykk9 committed:
[CBRD-24516] rev: replace the information which is printed in "show trace" from aggregate function optimization (#4370)

http://jira.cubrid.org/browse/CBRD-24516

Trace Statistics:
SELECT (time: 0, fetch: 12, ioread: 0)
SCAN (table: dba.t1), (aggregate optimized, time: 0, fetch: 0, ioread: 0, readrows: 0, rows: 0)
-->
SCAN (table: dba.t1), (noscan time: 0, fetch: 0, ioread: 0, readrows: 0, rows: 0, agl:pk_t1_id)

youngjinj commented on src/parser/csql_grammar.y:
```diff
@@ -21587,45 +21557,129 @@ opt_encrypt_algorithm
+            container_3 ctn;
+            memset(&ctn, 0x00, sizeof(ctn));
+            ctn.c1 = 0; // not online
+            ctn.c2 = prm_get_integer_value (PRM_ID_DEDUPLICATE_KEY_LEVEL);
+            ctn.c3 = prm_get_integer_value (PRM_ID_DEDUPLICATE_MIN_KEYS);
```
The result below seems to be the unintended result.
```
$ csql -u dba demodb -S

csql> ;edit
drop table if exists t1;
create table t1 (c1 int, c2 int, c3 int, c4 int, c5 int);
set system parameters 'deduplicate_key_level=11';
create index i1_level11 on t1 (c1);
set system parameters 'deduplicate_key_level=12';
create index i2_level12 on t1 (c2);
set system parameters 'deduplicate_key_level=13';
create index i3_level13 on t1 (c3);
set system parameters 'deduplicate_key_level=14';
create index i4_level14 on t1 (c4);
set system parameters 'deduplicate_key_level=15';
create index i5_level15 on t1 (c5);

csql> ;x
csql> ;sc t1
```
Expected Result
```
 <Constraints>

     INDEX i1_level11 ON dba.t1 (c1) WITH DEDUPLICATE=11
     INDEX i2_level12 ON dba.t1 (c2) WITH DEDUPLICATE=12
     INDEX i3_level13 ON dba.t1 (c3) WITH DEDUPLICATE=13
     INDEX i4_level14 ON dba.t1 (c4) WITH DEDUPLICATE=14
     INDEX i5_level15 ON dba.t1 (c5) WITH DEDUPLICATE=15
```
Actual Result
```
 <Constraints>

     INDEX i1_level11 ON dba.t1 (c1) WITH DEDUPLICATE=10
     INDEX i2_level12 ON dba.t1 (c2) WITH DEDUPLICATE=10
     INDEX i3_level13 ON dba.t1 (c3) WITH DEDUPLICATE=10
     INDEX i4_level14 ON dba.t1 (c4) WITH DEDUPLICATE=10
     INDEX i5_level15 ON dba.t1 (c5) WITH DEDUPLICATE=10
```

ctshim replied:
Good check.
I will fix it.  thanks.

ctshim committed:
trivia, adopt PR reviewer's comment

youngjinj committed:
[CBRD-24815] The precision of the NUMERIC type is set differently for SMALLINT, INT, and BIGINT types (#4375)

http://jira.cubrid.org/browse/CBRD-24815

This is about how to set the precision and scale when changing to NUMERIC type. 

1. Precision
   - AS-IS
      ```
      /* src/parser/type_checking.c - pt_coerce_expression_argument */

      bigint, int, smallint -> numeric (15, ?) /* DB_DEFAULT_NUMERIC_PRECISION */ 
      ```
   - TO-BE
      ```
      /* src/parser/type_checking.c - pt_coerce_expression_argument */

      bigint   -> numeric (19, ?) /* DB_BIGINT_PRECISION */
      int      -> numeric (10, ?) /* DB_INTEGER_PRECISION */
      smallint -> numeric (5, ?)  /* DB_SMALLINT_PRECISION */
      ```
2. Use macro variables instead of magic numbers.

ctshim committed:
Delay reading of system parameter settings after parsing.

beyondykk9 committed:
[CBRD-24684] divide operation result into numeric type (#4150)

http://jira.cubrid.org/browse/CBRD-24684

ctshim committed:
fix conflict

ctshim committed:
fix indent
